[JIT] add support for overloading functions

eellison opened the pull request:
This is a series of PRs that will allow us to support adding [padding to conv](https://github.com/pytorch/pytorch/pull/22484) and also reduce the friction of adding method overloads that was brought up in  https://github.com/pytorch/pytorch/pull/23266.

Support for overloaded functions following the specification in [PEP 484](https://www.python.org/dev/peps/pep-0484/#function-method-overloading).

The usage is:
```
@torch.jit.overload
def add(x: int, y: int) -> int: ...
@torch.jit.overload
def add(x: float, y: float) -> float: ...

def add:
    return x + y
```

Follow up PRs:

- Add same API for methods
- A couple of cleanups for functions:
     - don't require default params specified on the overload as well
     - potentially error if invocation could be matched to multiple overloads. now it just chooses the first one, mypy does the same thing currently
 

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23886 [JIT] add support for overloading functions**
* #23885 [JIT] metacompile isinstance checks

Differential Revision: [D16694863](https://our.internmc.facebook.com/intern/diff/D16694863)

Commit:
add support for overloading functions

Commit:
Update on "[JIT] add support for overloading functions"

This is a series of PRs that will allow us to support adding [padding to conv](https://github.com/pytorch/pytorch/pull/22484) and also reduce the friction of adding method overloads that was brought up in  https://github.com/pytorch/pytorch/pull/23266.

driazati commented on torch/jit/__init__.py:
```diff
@@ -2079,6 +2079,80 @@ def _get_script_class(name):
+def overload(func):
+    qual_name = _qualified_name(func)
+    global _overloaded_fns
+    li = _overloaded_fns.get(qual_name)
```
Can you use a more descriptive name for `li`

driazati commented on torch/jit/__init__.py:
```diff
@@ -2079,6 +2079,80 @@ def _get_script_class(name):
+def compile_function_with_overload(qual_name, impl_fn, overload_decl, overload_defaults):
+    _frames_up = 0
+    impl_ast = torch.jit.get_jit_def(impl_fn)
+    # TODO: figure out correct closure_rcb, stack_rcb, although this seems to work
```
You probably need to leave this as is, see the comment in `_rcb` below, but we should at least avoid copy pasting this code to make a rcb from `script`

driazati commented on torch/jit/__init__.py:
```diff
@@ -2079,6 +2079,80 @@ def _get_script_class(name):
+        compiled_fns.append(compiled_fn)
+
+    # cache compilation, remove information stored to do compilation
+    _compiled_overloaded_fns[qual_name] = compiled_fns
```
This looks like a potential memory leak, if the functions this compiles go out of scope, their compiled versions will stick around here. @Chillee found a similar issue recently, I don't remember if he found a fix or not

eellison replied:
Discussed in person: this is not a public api yet, and will have very limited usage here so we are going to leave it as is for now.

eellison commented on test/test_jit.py:
```diff
@@ -12832,6 +12832,91 @@ def test_non_primitive_types(x):
         self.assertEqual(out, torch.tensor(6.0))
 
+    def test_function_overloads(self):
+        # TODO: pyflakes currently does not compose @overload annotation with other
```
Will do when this lands

Commit:
Update on "[JIT] add support for overloading functions"

This is a series of PRs that will allow us to support adding [padding to conv](https://github.com/pytorch/pytorch/pull/22484) and also reduce the friction of adding method overloads that was brought up in  https://github.com/pytorch/pytorch/pull/23266.

Support for overloaded functions following the specification in [PEP 484](https://www.python.org/dev/peps/pep-0484/#function-method-overloading).

The usage is:
```
@torch.jit.overload
def add(x: int, y: int) -> int: ...
@torch.jit.overload
def add(x: int, y: int) -> int: ...

def add:
    return x + y
```

Follow up PRs:

- Add same API for methods
- A couple of cleanups for functions:
     - don't require default params specified on the overload as well
     - potentially error if invocation could be matched to multiple overloads. now it just chooses the first one, mypy does the same thing currently

Commit:
Update on "[JIT] add support for overloading functions"

This is a series of PRs that will allow us to support adding [padding to conv](https://github.com/pytorch/pytorch/pull/22484) and also reduce the friction of adding method overloads that was brought up in  https://github.com/pytorch/pytorch/pull/23266.

Support for overloaded functions following the specification in [PEP 484](https://www.python.org/dev/peps/pep-0484/#function-method-overloading).

The usage is:
```
@torch.jit.overload
def add(x: int, y: int) -> int: ...
@torch.jit.overload
def add(x: int, y: int) -> int: ...

def add:
    return x + y
```

Follow up PRs:

- Add same API for methods
- A couple of cleanups for functions:
     - don't require default params specified on the overload as well
     - potentially error if invocation could be matched to multiple overloads. now it just chooses the first one, mypy does the same thing currently

Commit:
Update on "[JIT] add support for overloading functions"

This is a series of PRs that will allow us to support adding [padding to conv](https://github.com/pytorch/pytorch/pull/22484) and also reduce the friction of adding method overloads that was brought up in  https://github.com/pytorch/pytorch/pull/23266.

Support for overloaded functions following the specification in [PEP 484](https://www.python.org/dev/peps/pep-0484/#function-method-overloading).

The usage is:
```
@torch.jit.overload
def add(x: int, y: int) -> int: ...
@torch.jit.overload
def add(x: int, y: int) -> int: ...

def add:
    return x + y
```

Follow up PRs:

- Add same API for methods
- A couple of cleanups for functions:
     - don't require default params specified on the overload as well
     - potentially error if invocation could be matched to multiple overloads. now it just chooses the first one, mypy does the same thing currently

Chillee commented on torch/jit/__init__.py:
```diff
@@ -2079,6 +2085,71 @@ def _get_script_class(name):
+    if qual_name in _compiled_overloaded_fns or qual_name in _overloaded_fns:
+        raise RuntimeError("Function {} cannot be directly compiled because it"
+                           " is overloaded. Please wrap it with typed inputs of"
+                           " the signature you wish to compile instead".format(obj))
```
I think this should be `qual_name` right?

Chillee commented:
Also, can you add a full example somewhere? I'm having trouble getting this to work.

I have 
```
@torch.jit.overload  # noqa: F811
def identity(x1):  # noqa: F811
    # type: (str) -> str
    pass

@torch.jit.overload  # noqa: F811
def identity(x1=1.0):  # noqa: F811
    # type: (float) -> float
    pass

def identity(x1=1.0):  # noqa: F811
    return x1

print(torch.jit.script(identity))
```

This throws an error:
```
RuntimeError: Function <function identity at 0x7fba94ef41e0> cannot be directly compiled because it is overloaded. Please wrap it with typed inputs of the signature you wish to compile instead
```

Commit:
Update on "[JIT] add support for overloading functions"

This is a series of PRs that will allow us to support adding [padding to conv](https://github.com/pytorch/pytorch/pull/22484) and also reduce the friction of adding method overloads that was brought up in  https://github.com/pytorch/pytorch/pull/23266.

Support for overloaded functions following the specification in [PEP 484](https://www.python.org/dev/peps/pep-0484/#function-method-overloading).

The usage is:
```
@torch.jit.overload
def add(x: int, y: int) -> int: ...
@torch.jit.overload
def add(x: float, y: float) -> float: ...

def add:
    return x + y
```

Follow up PRs:

- Add same API for methods
- A couple of cleanups for functions:
     - don't require default params specified on the overload as well
     - potentially error if invocation could be matched to multiple overloads. now it just chooses the first one, mypy does the same thing currently

eellison commented:
> Also, can you add a full example somewhere? I'm having trouble getting this to work.
> 
> I have
> 
> ```
> @torch.jit.overload  # noqa: F811
> def identity(x1):  # noqa: F811
>     # type: (str) -> str
>     pass
> 
> @torch.jit.overload  # noqa: F811
> def identity(x1=1.0):  # noqa: F811
>     # type: (float) -> float
>     pass
> 
> def identity(x1=1.0):  # noqa: F811
>     return x1
> 
> print(torch.jit.script(identity))
> ```
> 
> This throws an error:
> 
> ```
> RuntimeError: Function <function identity at 0x7fba94ef41e0> cannot be directly compiled because it is overloaded. Please wrap it with typed inputs of the signature you wish to compile instead
> ```
> 
> Specifically, it's not clear to me what "wrapping it with typed inputs" means.
> 
> EDIT: I figured it out by looking at the screenshot you initially sent me. I'd suggest rephrasing the error as something like:
> 
> ```
> RuntimeError: Function <qualname> cannot be directly compiled because it is overloaded. It must be used in the context of a function where `isinstance` checks can statically determine the type.
> ```
> 
> Perhaps less technically involved. An example might even be a good idea.

Thanks for the comments. Updated the error message, let me know if you think it's easier to understand. Btw, I suspect this is a very uncommon case (compiling a single function from nn/functional by itself).

driazati commented on test/test_jit.py:
```diff
@@ -12832,6 +12832,113 @@ def test_non_primitive_types(x):
+        def schema_match_failure():
+            return identity((1, 2))
+
+        thrown = False
```
Why not `self.assertRaisesRegex`?

eellison replied:
Couldn't figure out the regex lol

Chillee replied:
Would look something like this if you wanna do it with regex (gotta use lookaheads though): https://regexr.com/4irtv

`(?=.*of type 'str')(?=.*of type 'float').*`

Might not be worth it.

driazati commented on test/test_jit.py:
```diff
@@ -12832,6 +12832,113 @@ def test_non_primitive_types(x):
+
+        def my_conv(x, y=2.0):  # noqa: F811
+            if isinstance(y, str):
+                return 4.0 - x
```
Can you also test the `str` path here?

driazati commented on torch/csrc/jit/script/init.cpp:
```diff
@@ -173,6 +173,24 @@ std::shared_ptr<PythonResolver> pythonResolver(
+  const auto& new_params = new_decl.params();
+  const auto& old_params = old_decl.params();
+
+  TORCH_INTERNAL_ASSERT(
```
It doesn't really matter for our use case but add a comment here that this isn't strictly necessary, it looks like the overloads can have [different signatures](https://mypy.readthedocs.io/en/latest/more_types.html#function-overloading)

eellison replied:
Yup, this is along the same path as not requiring default params on the overload that is a follow up

driazati commented on torch/csrc/jit/script/init.cpp:
```diff
@@ -702,6 +720,34 @@ void initJitScriptBindings(PyObject* module) {
+        const auto name = c10::QualifiedName(qualname);
+        auto cu = get_python_cu();
+        checkOverloadDecl(overload_decl, implementation_def.decl());
+        auto new_def = implementation_def.withDecl(overload_decl);
```
It looks like replacing the `implementation_def` is all this does differently from `_jit_script_compile`, that should be moved to a function so we don't have to copy paste it here

driazati commented on torch/jit/__init__.py:
```diff
@@ -2079,6 +2085,68 @@ def _get_script_class(name):
+    fn_overload_list.append((torch.jit.get_jit_def(func).decl(), get_default_args(func)))
+    return func
+
+def compile_function_with_overload(qual_name, impl_fn, overload_decl, overload_defaults):
```
These would all appear in `torch.jit`, so prefix everything but `overload` with a `_`

driazati reviewed (approved):
Looks good, remaining stuff is mostly just structural

Commit:
Update on "[JIT] add support for overloading functions"

This is a series of PRs that will allow us to support adding [padding to conv](https://github.com/pytorch/pytorch/pull/22484) and also reduce the friction of adding method overloads that was brought up in  https://github.com/pytorch/pytorch/pull/23266.

Support for overloaded functions following the specification in [PEP 484](https://www.python.org/dev/peps/pep-0484/#function-method-overloading).

The usage is:
```
@torch.jit.overload
def add(x: int, y: int) -> int: ...
@torch.jit.overload
def add(x: float, y: float) -> float: ...

def add:
    return x + y
```

Follow up PRs:

- Add same API for methods
- A couple of cleanups for functions:
     - don't require default params specified on the overload as well
     - potentially error if invocation could be matched to multiple overloads. now it just chooses the first one, mypy does the same thing currently
 


Differential Revision: [D16694863](https://our.internmc.facebook.com/intern/diff/D16694863)

Commit:
Update on "[JIT] add support for overloading functions"

This is a series of PRs that will allow us to support adding [padding to conv](https://github.com/pytorch/pytorch/pull/22484) and also reduce the friction of adding method overloads that was brought up in  https://github.com/pytorch/pytorch/pull/23266.

Support for overloaded functions following the specification in [PEP 484](https://www.python.org/dev/peps/pep-0484/#function-method-overloading).

The usage is:
```
@torch.jit.overload
def add(x: int, y: int) -> int: ...
@torch.jit.overload
def add(x: float, y: float) -> float: ...

def add:
    return x + y
```

Follow up PRs:

- Add same API for methods
- A couple of cleanups for functions:
     - don't require default params specified on the overload as well
     - potentially error if invocation could be matched to multiple overloads. now it just chooses the first one, mypy does the same thing currently
 


Differential Revision: [D16694863](https://our.internmc.facebook.com/intern/diff/D16694863)

Commit:
Update on "[JIT] add support for overloading functions"

This is a series of PRs that will allow us to support adding [padding to conv](https://github.com/pytorch/pytorch/pull/22484) and also reduce the friction of adding method overloads that was brought up in  https://github.com/pytorch/pytorch/pull/23266.

Support for overloaded functions following the specification in [PEP 484](https://www.python.org/dev/peps/pep-0484/#function-method-overloading).

The usage is:
```
@torch.jit.overload
def add(x: int, y: int) -> int: ...
@torch.jit.overload
def add(x: float, y: float) -> float: ...

def add:
    return x + y
```

Follow up PRs:

- Add same API for methods
- A couple of cleanups for functions:
     - don't require default params specified on the overload as well
     - potentially error if invocation could be matched to multiple overloads. now it just chooses the first one, mypy does the same thing currently
 


Differential Revision: [D16694863](https://our.internmc.facebook.com/intern/diff/D16694863)

suo commented on torch/csrc/jit/script/python_sugared_value.cpp:
```diff
@@ -544,6 +573,13 @@ std::shared_ptr<SugaredValue> toSugaredValue(
 
   py::bool_ isFunction = py::module::import("inspect").attr("isfunction")(obj);
   if (py::cast<bool>(isFunction)) {
+    auto overloads =
```
Looks like the infrastructure for overloaded functions is only implemented for Python. Do we have a plan for resolving overloaded functions defined from the string frontend (for example, how will we serialize overloaded functions)?

eellison replied:
Any overloaded function that needs to be serialized will already be instantiated in one of its overloads, that function gets serialized normally.

suo commented on torch/jit/__init__.py:
```diff
@@ -2110,6 +2116,68 @@ def _get_script_class(name):
+    qual_name = _qualified_name(obj)
+    global _compiled_overloaded_fns
+    compiled_overloads = _compiled_overloaded_fns.get(qual_name, None)
+    if compiled_overloads is not None:
```
This doesn't work if we call `_get_overloads()` then add another overload after. The second overload will not ever get compiled. Do we need a cache here? We don't have anything similar for other recursively compiled functions.

eellison replied:
What is the use case you are imagining? This is still an internal only api so I don't think it applies but something to think about if we make it public. We don't strictly need a cache it's true.

suo replied:
If we add a new overload after using the overload in the scripted function, it will not be registered. Let's remove the cache, or make the existing caching logic correct. Even if it's unlikely I think it'll be very surprising if we ever run into this.

suo commented on torch/csrc/jit/script/init.cpp:
```diff
@@ -423,8 +463,8 @@ void initJitScriptBindings(PyObject* module) {
           [](Module& self, std::string name, TypePtr type, py::object value) {
-            auto unshaped = unshapedType(type);
-            self.register_attribute(name, unshaped, toIValue(value, type));
+            self.register_attribute(
```
Why was this changed?

eellison replied:
Clang tidy complained
